Add testrunner design documentation (#687)
SUSE/skuba: ci/infra/testrunner/README.md

Testrunner is a CLI tool for setting up an environment for running e2e tests, making transparent the mechanism used for providing the test infrastructure. It can be used as a stand-alone tool, running either locally or as part of a CI pipeline. It provides commands for provisioning the infrastructure, deploying a k8s cluster using `skuba`,  and running tests. It also provides [a library](tests/README.md) for developing `pytest`-based tests.

## Design

The `testrunner` is composed by multiple components as described in the figure below:

> * The `testrunner.py` frontend cli application, which receives CLI options, loads a configuration and executes the selected command. The configuration is loaded from a yaml file (by default, [vars.yaml](vars.yaml)) and mixed with the environment variables.

Reviewer: Inconsistent cli v/s CLI capitalization
```suggestion
* The `testrunner.py` frontend CLI application, which receives CLI options, loads a configuration and executes the selected command. The configuration is loaded from a yaml file (by default, [vars.yaml](vars.yaml)) and mixed with the environment variables.
```

> * The `testrunner.py` frontend cli application, which receives CLI options, loads a configuration and executes the selected command. The configuration is loaded from a yaml file (by default, [vars.yaml](vars.yaml)) and initialized by the `BaseConfig` class (located in [utils/constanst.py](utils/constanst.py) which mixes it with the environment variables and sets defaults values for missing configuration parameters.   

Reviewer 2: Do you think it provides value to go that detailed about the `BaseConfig` class?

Author: In practice I've found most people is not aware of this class and how it works, mostly related to how it sets defaults and merge configuration. I actually would like to rename the module to `config.py` to make more evident its intention.

Author: I'm removing this detail, but I think it should be documented somewhere, maybe a section on how to add new configuration parameters.

* A set of supporting classes which offer functionality for setting up and interacting with a cluster and executing tests. These classes wrap external tools such as `skuba`, `terraform`, `kubectl` and `pytests`. The commands offered by the `testrunner.py` cli application mostly expose the functionalities of these classes and add only the required glue code.

> * In the case of the test command, the `testrunner` does not only wrap the `pytest` testing tool ([tests/driver.py](tests/driver.py), but also offers a test library ([tests/conftest.py](tests/conftest.py)) which implements reusable test functionalities as well as `fixtures` to facilitate [test development](tests/README.md). These fixtures use the libraries offered by testrunner for setting up the test infrastructure, deploying the cluster, interact with the cluster and the nodes, cleaning up after execution, execute common validation checks, among others.

Reviewer: ```suggestion
* In the case of the test command, the `testrunner` not only wraps the `pytest` testing tool ([tests/driver.py](tests/driver.py), but also offers a test library ([tests/conftest.py](tests/conftest.py)) which implements reusable test functions as well as `fixtures` to facilitate [test development](tests/README.md). These fixtures use the libraries offered by testrunner for setting up the test infrastructure, deploying the cluster, interacting with the cluster and the nodes, cleaning up after execution, and executing common validation checks among others.
```
Minor grammar agreement

Author: I had like this and @ereslibre suggested how it is now  :sweat_smile:

Reviewer 2: Sorry, ugh.

> * In the case of the test command, the `testrunner` not only wrap the `pytest` testing tool ([tests/driver.py](tests/driver.py), but also offers a test library ([tests/conftest.py](tests/conftest.py)) which implements reusable test functionalities or `fixtures` to facilitate [test development](tests/README.md). These fixtures use the libraries offered by testrunner for setting up the test infrastructure, deploying the cluster, interact with the cluster and the nodes, cleaning up after execution, execute common validation checks, among others.

Reviewer 2: ```suggestion
* In the case of the test command, the `testrunner` does not only wrap the `pytest` testing tool ([tests/driver.py](tests/driver.py), but also offers a test library ([tests/conftest.py](tests/conftest.py)) which implements reusable test functionalities as well as `fixtures` to facilitate [test development](tests/README.md). These fixtures use the libraries offered by testrunner for setting up the test infrastructure, deploying the cluster, interact with the cluster and the nodes, cleaning up after execution, execute common validation checks, among others.
```

The objective of this design is to maintain a clear separation between the following concerns:
* User interface including configuration (`testrunner.py` and `BaseConfig` class)
* Wrapping external classes (supporting libraries for `skuba`, `terraform`, `kubectl`)
* Reusable test functionalities (`conftest.py`)
* Test logic (diverse test suites)

> It is worth noticing that tests can be executed directly using pytest, but it is more convenient to execute them using the testrunner, resulting in a consistent user experience.

Reviewer: ```suggestion
It is worth noting that, although tests can be executed directly using pytest, executing them using the testrunner is more convenient and results in a consistent user experience.
```
Minor rewording to reduce commas

> It worth noticing that tests can be executed directly using pytest, but it results convenient to execute them by means of the testrunner a consistent user experience. 

Reviewer 2: ```suggestion
It is worth noticing that tests can be executed directly using pytest, but it is more convenient to execute them using testrunner, resulting in a consistent user experience. 
```

```                                           
                                              | env variables
                                              v
              +--------------+         +----------------+     +---------------+
  CLI options |              |<--------|                |     | Configuration |

>   ----------->|  Testrunner  |         | Initialiation  |<----|  (vars.yaml)  |

Reviewer 2: ```suggestion
  ----------->|  Testrunner  |         | Initialization  |<----|  (vars.yaml)  |
```
